### PR TITLE
CASM-4669 Include image signatures in csm tarball

### PR DESCRIPTION
**Pull request by mtupitsyn**
## Summary and Scope

We already sign images during build with one of 3 keys:
* Images built with Github actions: `gcpkms://projects/sdlc-ops/locations/global/keyRings/csm-builds/cryptoKeys/github-cray-hpe/versions/1`
* Images built with Jenkins: `gcpkms://projects/sdlc-ops/locations/global/keyRings/csm-builds/cryptoKeys/jenkins-csm/versions/1`
* New GCP key infrastructure, established with CASM-4667 (stage only ATM, prod instance planned): `gcpkms://projects/hpe-stage-csm-security/locations/global/keyRings/cosign/cryptoKeys/csm-images/versions/1`

This PR provides the following changes:

#### During CSM tarball packaging:
* Image signatures are verified with provided set of keys
* Images packaged with signatures included
* Public keys are packaged

#### During CSM install / upgrade
* Images are uploaded into PIT Nexus and Kubernetes Nexus together with respective signatures
* Public keys are injected into Helm customizations under `spec.kubernetes.services.kyverno-policy.checkImagePolicy.publicKeys` key - so that they can be consumed by `kyverno-policy` chart in this manner: https://github.com/Cray-HPE/cray-kyverno-policies/blob/feature/image-signatures/charts/kyverno-policy/templates/cluster/cluster-check-image.yaml#L29-L36

## Issues and Related PRs

* Resolves CASM-4669
* Relates to CASM-4667

### Tested on:

  * Virtual Shasta
  * Jenkins build
  * Local testing

#### During build:

  * Tested that all images participating build are properly signed, by executing new stage during build: 
image
  * Performed false positive test - injected unsigned image into build, ensured that `cosign verify` step failed

#### After installing temporary build on vShasta system, ensured that:

  * Images are available in both PIT Nexus and Cloud Nexus together with signatures: 
image
  * Kyverno policy chart, updated to consume public keys from customization, can be installed with proper public keys configuration, which is provided via customizations:

        ncn-m001:~ # helm -n kyverno get values kyverno-policy
        USER-SUPPLIED VALUES:
        checkImagePolicy:
          publicKeys:
          - |-
            -----BEGIN PUBLIC KEY-----
            MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAERxLYOVq/kBeE05mUZqTk85CaOpSC
            CdYeLlCt+K941eQgNWQLBdMiDPnicw5i9o278apo/OLD5AZVSX8ZXPYkKQ==
            -----END PUBLIC KEY-----
          - |-
            -----BEGIN PUBLIC KEY-----
            MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEhEQ9X1j2d7qDHEXJRbYnAcYMhGop
            OBhknIvSWQYhOpiH74CtJX4/KKPOQJxmb+ZHdZjC9GQZzxyayp2EvLF2Og==
            -----END PUBLIC KEY-----
          - |-
            -----BEGIN PUBLIC KEY-----
            MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEArrdSOxAr5Pn/WgxSvrcU
            ufjL2aGknFOPeNSkQiI5s4etVlWz5Om21zgBSgatf4PrKPSv4wjG3ujKZqSJ6o8k
            jTO/ycNcsJnkqRROS158tOwGeMy5OQQaH4n+0tEYFvRp3asVqCe8jU4jsAZw3+CY
            eIGMgzbNVoZlvjCl9DQQeMBqrjLefsqUUcc2HwL+PiBIi8pzhGvWfgDpHMkH4GeU
            vdfVkKGW27azhUFj9HbI6njDIvpNaGIBh47hnM+dAdi/9/c/Rs6nx2rpaq3OPR19
            2fxNN7rvReJIE8Ce5HMccp62ihS8F5WGDMlNdmZ2BR8h/DIoSxZ1iySUD1gqlTnb
            EVLJrMU5m/yS3fVAxG2oub29bcvj4r4qwmpBXQT1impoqw4wQZO7nFPOUfW9SIXW
            668sN6khTRbPQBXB6BBZYOTVibpXxl61D9WuBf/bs7epxbViFr2DgxGfebYb+wlw
            AfPzoJ5gX6f0r7Ob5PK7exLbf66XWClxvZ9nr6EYN3x3IQCdYDa2pcLU59wBygzM
            5qBsTOuePJnq+MeYWoJmSmodah5NxLkZjIPBncz5NPWgJC0sI/pTx4X/8Uh7w74D
            MhH5NnfxQu5FQQ1bSyicbH0pJAgPITflcMvyeKU0pZgLKJRi6xUHMFm6tQIsj2Y8
            +KI4ZSRDM5psovIlaqOm+hUCAwEAAQ==
            -----END PUBLIC KEY-----
        global:
          chart:
            name: kyverno-policy
            version: 1.4.3

> [!WARNING]
> I noticed that Kyverno seems to have issues with verification with public keys in our system, at least in version 1.9.5. It does not support concept of mirroring, so it tries to download image manifests directly from `artifactory.algol60.net`, rather then trying mirrors configured in `/etc/containerd/config.toml` on Kubernetes nodes. Also, even with `failurePolicy` set to `Ignore` and `validationFailureAction` set to `Audit` Kyverno seems to be blocking some network communication during pod deployment.

#### Performed upgrade from last 1.5.1 build to temporary build on vShasta:
https://jenkins.algol60.net/blue/organizations/jenkins/Cray-HPE%252Fcsm-vshasta-deploy/detail/feature%252Fimage-signatures/9/pipeline

## Risks and Mitigations

Low - new feature

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: Jenkinsfile.github [hunk 1]
```diff
@@ -47,8 +47,10 @@ pipeline {
         ARTIFACTORY = credentials('artifactory-algol60-readonly')
         PARALLEL_JOBS = "75%"
         SNYK_TOKEN = credentials('SNYK_TOKEN')
-        SLACK_CHANNEL_NOTIFY = "${isPublishedRelease() ? "#casm_release_management" : ""}"
-        SLACK_CHANNEL_ALERTS = "${isPublishedRelease() ? "#csm-release-alerts" : ""}"
+        // SLACK_CHANNEL_NOTIFY = "${isPublishedRelease() ? "#casm_release_management" : ""}"
+        // SLACK_CHANNEL_ALERTS = "${isPublishedRelease() ? "#csm-release-alerts" : ""}"
+        SLACK_CHANNEL_NOTIFY = "${env.TAG_NAME ? "@mikhail.tupitsyn" : ""}"
+        SLACK_CHANNEL_ALERTS = ""
```
Comment (mtupitsyn) [inline Jenkinsfile.github]: Temporary change for testing.

File: common.sh [hunk 2]
```diff
@@ -25,8 +25,8 @@
 
 export PACKAGING_TOOLS_IMAGE=${PACKAGING_TOOLS_IMAGE:-artifactory.algol60.net/dst-docker-mirror/internal-docker-stable-local/packaging-tools:0.14.0}
 export RPM_TOOLS_IMAGE=${RPM_TOOLS_IMAGE:-artifactory.algol60.net/dst-docker-mirror/internal-docker-stable-local/rpm-tools:1.0.0}
-export SKOPEO_IMAGE=${SKOPEO_IMAGE:-artifactory.algol60.net/dst-docker-mirror/quay-remote/skopeo/stable:v1.13.2}
-export CRAY_NEXUS_SETUP_IMAGE=${CRAY_NEXUS_SETUP_IMAGE:-artifactory.algol60.net/csm-docker/stable/cray-nexus-setup:0.7.1}
+export SKOPEO_IMAGE=${SKOPEO_IMAGE:-artifactory.algol60.net/csm-docker/stable/quay.io/skopeo/stable:v1}
```
Comment (mtupitsyn) [inline common.sh:28]: A skopeo image which we have under `artifactory.algol60.net/csm-docker/stable/quay.io/skopeo/stable:v1` has additional configuration setting, instructing it to copy images together with Sigstore attachments.

File: manifests/nexus.yaml [hunk 3]
```diff
@@ -10,5 +33,5 @@ spec:
   charts:
   - name: cray-nexus
     source: csm-algol60
-    version: 0.11.1
+    version: 0.12.1
```
Comment (mtupitsyn) [inline manifests/nexus.yaml:36]: To store images together with Sigstore attachments in Cloud Nexus, we have to upgrade Nexus to 3.67.1 (and consequently upgrade `cray-nexus-setup` image to `0.11.0`).

**Review by rustydb — APPROVED** [review]

**Comment by mohansunderv-hpe** [comment]
approved

**Pull request merged**